PyInstaller/building/build_main.py

```diff
@@ -109,29 +108,21 @@ def discover_hook_directories():
     :return: list of discovered hook directories.
     """
 
-    hook_directories = []
-    output = exec_statement("""
-        import sys
-        import pkg_resources
+    import sys
+    import pkg_resources
+    from PyInstaller.log import logger
 
-        entry_points = pkg_resources.iter_entry_points(
-            'pyinstaller40', 'hook-dirs')
-        for entry_point in entry_points:
-            try:
-                hook_dirs = entry_point.load()()
-                for hook_dir in hook_dirs:
-                    print('\\n$_pyi:' + hook_dir + '*')
-            except Exception as e:
-                print("discover_hook_directories: Failed to process hook "
-                      "entry point '%s': %s" % (entry_point, e),
-                      file=sys.stderr)
-    """)
-
-    for line in output.split():
-        # Filter out extra output by checking for the special prefix
-        # and suffix
-        if line.startswith("$_pyi:") and line.endswith("*"):
-            hook_directories.append(line[6:-1])
+    entry_points = pkg_resources.iter_entry_points(
+        'pyinstaller40', 'hook-dirs')
+
+    hook_directories = []
+    for entry_point in entry_points:
+        try:
+            hook_directories.extend(entry_point.load()())
+        except Exception as e:
+            print("discover_hook_directories: Failed to process hook "
+                  "entry point '%s': %s" % (entry_point, e),
+                  file=sys.stderr)
```

Comment:
I suppose this `print()` should be changed to `logger.warning()` if we are going to use `PyInstaller.log.logger` within subprocess.

Reply:
If we are to go down the path of using `PyInstaller.log.logger` in subprocesses then perhaps it would be better to define the logger globally in the child and provide it [here](https://github.com/pyinstaller/pyinstaller/pull/6052/commits/50314d96aff9e364c444ea3c101ed1b1b08ba001#diff-f56f61caa4463d116def5b7e5e6de14869928d4fe1409a4422f7d1a04d096bd4R54).

Reply:
> I was going to ask about this. Technically it's incorrect to pull in another submodule's logger because each logger is supposed to be tied to the submodule it was defined in

This is true if logger is created as `logger = logging.getLogger(__name__)`, because __name__ "ties" it to the module (i.e., the instance has module's name, and can be configured independently of other modules' loggers - that is, if we used log config file).

But it looks like `PyInstaller.log.logger` is a "global" logger, with shared name ("PyInstaller"):
https://github.com/pyinstaller/pyinstaller/blob/fee253f15946b31aab629e12ed135a81f730648f/PyInstaller/log.py#L27

The module names (where module-named instance is used) not showing up is due to our custom format not including that piece of info:
https://github.com/pyinstaller/pyinstaller/blob/fee253f15946b31aab629e12ed135a81f730648f/PyInstaller/log.py#L25

But I think the rest of code in `PyInstaller.log` implies that `--log-level` command-line switch applies only to the global "PyInstaller" logger but not to the module-named ones (if there are any).

Reply:
> But I think the rest of code in `PyInstaller.log` implies that `--log-level` command-line switch applies only to the global "PyInstaller" logger but not to the module-named ones (if there are any).

Or it might actually apply to them all, because module-named ones also start with the `PyInstaller.` prefix.

Reply:
> If we are to go down the path of using `PyInstaller.log.logger` in subprocesses then perhaps it would be better to define the logger globally in the child and provide it [here](https://github.com/pyinstaller/pyinstaller/pull/6052/commits/50314d96aff9e364c444ea3c101ed1b1b08ba001#diff-f56f61caa4463d116def5b7e5e6de14869928d4fe1409a4422f7d1a04d096bd4R54)?

That will obfuscate the origin of `logger` within `@isolated.decorate`'d function, which I think will make those function harder to read, as it will seem that `logger` is coming from parent module's namespace; or not being defined anywhere, which might also upset the linter...

Reply:
Good, I didn't like the idea either but I thought I'd bounce it around anyway.